### PR TITLE
expand: avoid panics on division by zero

### LINKED ISSUE #892: interp: panic on division by zero
Given the invalid command 
```
let x/x
``` 
bash rightfully complains:
```
$ let x/x
bash: let: x/x: division by 0 (error token is "x")
``` 
but gosh panics:
```
$ go run cmd/gosh/main.go
$ let x/x
panic: runtime error: integer divide by zero

goroutine 1 [running]:
mvdan.cc/sh/v3/expand.binArit(0x111540?, 0x589de0?, 0xc000122f28?)
	/home/dank/src/sh/expand/arith.go:173 +0x245
mvdan.cc/sh/v3/expand.Arithm(0x0?, {0x589840?, 0xc000078480?})
	/home/dank/src/sh/expand/arith.go:95 +0x353
...
```
This happens in v3.0.0 and master. I didn't try any tags older than v3.0.0.  Tested with go 1.18.1.

### PR DESCRIPTION
**Pull request by mvdan**
(see commit message)

Fixes #892.

### CODE REVIEW

**Comment by mvdan** [comment]
cc @dankegel

**Review by andreynering — APPROVED** [review]

File: expand/arith.go [hunk 1]
```diff
@@ -161,48 +164,51 @@ func intPow(a, b int) int {
...
+		return x * y, nil
 	case syntax.Quo:
-		return x / y
+		if y == 0 {
+			return 0, fmt.Errorf("division by zero")
+		}
+		return x / y, nil
 	case syntax.Rem:
-		return x % y
+		return x % y, nil
```
Comment (theclapp) [inline expand/arith.go:181]: `x % 0` can also panic.

File: interp/api.go [hunk 2]
```diff
@@ -4,6 +4,13 @@
 // Package interp implements an interpreter that executes shell
 // programs. It aims to support POSIX, but its support is not complete
 // yet. It also supports some Bash features.
+//
+// The interpreter generally aims to behave like Bash,
+// but it does not support all of its features.
+//
+// The interpreter currently aims to behave like a non-interactive shell,
+// which how most shells run scripts, and is more useful to machines.
```
Comment (theclapp) [inline interp/api.go:12]: "which is how"

File: interp/handler_test.go [hunk 3]
```diff
@@ -21,29 +21,29 @@ import (
 	"mvdan.cc/sh/v3/syntax"
 )
 
-func blacklistBuiltinExec(name string) ExecHandlerFunc {
+func blocklistBuiltinExec(name string) ExecHandlerFunc {
```
Comment (theclapp) [inline interp/handler_test.go:24]: "blacklist" => "blocklist" 👍👍👍

File: expand/arith.go [hunk 4]
```diff
@@ -129,6 +129,9 @@ func (cfg *Config) assgnArit(b *syntax.BinaryArithm) (int, error) {
 	case syntax.MulAssgn:
 		val *= arg
 	case syntax.QuoAssgn:
+		if arg == 0 {
+			return 0, fmt.Errorf("division by zero")
+		}
 		val /= arg
 	case syntax.RemAssgn:
 		val %= arg
```
Comment (theclapp) [inline expand/arith.go:137]: `%= 0` can also panic.

File: interp/handler_test.go [hunk 5]
```diff
@@ -67,40 +67,40 @@ var modCases = []struct {
...
-		name: "ExecBlacklist",
-		exec: blacklistBuiltinExec("sleep"),
+		name: "Execblocklist",
+		exec: blocklistBuiltinExec("sleep"),
 		src:  "echo foo; sleep 1",
-		want: "foo\nsleep: blacklisted builtin",
+		want: "foo\nsleep: blocklisted builtin",
 	},
 	{
 		name: "ExecWhitelist",
```
Comment (theclapp) [inline interp/handler_test.go:76]: "blacklist" => "blocklist", "whitelist" => ?

File: interp/runner.go [hunk 6]
```diff
@@ -155,7 +156,19 @@ func (r *Runner) updateExpandOpts() {
...
-		r.errf("%v\n", err)
+		errMsg := err.Error()
+		fmt.Fprintln(r.stderr, errMsg)
+		switch {
+		case errors.As(err, &expand.UnsetParameterError{}):
+		case errMsg == "invalid indirect expansion":
+			// TODO: These errors are treated as fatal by bash.
+			// Make the error type reflect that.
+		case strings.HasSuffix(errMsg, "not supported"):
+			// TODO: This "contains" is a temporary measure until the expand
```
Comment (theclapp) [inline interp/runner.go:167]: You call `HasSuffix` but you comment about "contains", which could be exactly what you meant, but noting it if not.

**Review by theclapp — CHANGES_REQUESTED** [review]
Nice work.  Found some more panics with `%`, though.

File: interp/handler_test.go (hunk 3 above)
Comment (mvdan) [reply interp/handler_test.go:24]: Indeed on purpose, I just forgot to mention it in the commit message :) done.

File: interp/api.go (hunk 2 above)
Comment (mvdan) [reply interp/api.go:12]: done

File: interp/handler_test.go (hunk 5 above)
Comment (mvdan) [reply interp/handler_test.go:76]: forgot this one. it's not even an allowlist.

File: interp/runner.go (hunk 6 above)
Comment (mvdan) [reply interp/runner.go:167]: indeed a suffix on purpose - updated the code but not the comment.

File: expand/arith.go (hunk 1 above)
Comment (mvdan) [reply expand/arith.go:181]: Very well spotted. I actually looked into shifts, which we already were protected against, but I completely forgot about remainders. Done.

File: expand/arith.go (hunk 4 above)
Comment (mvdan) [reply expand/arith.go:137]: done

**Comment by mvdan** [comment]
Thanks for the great review, @theclapp. I'll probably bother you for more interpreter reviews, if that's okay :)

**Comment by mvdan** [comment]
Got a weird failure on Mac I haven't seen before. Maybe the machine crashed somehow.

**Review by theclapp — APPROVED** [review]

**Comment by mvdan** [comment]
I'm perplexed at the repeated macos failure. It happens on pull_request but not on push, which should be the same code. And master passed. And the failure is in the shfmt tests, where it depends on neither of the packages modified here. I'll rerun all the jobs and see...

**Comment by mvdan** [comment]
Green, so I will ignore that. For the record, they were `signal: killed` errors in the `cmd/shfmt` test scripts.

**Pull request merged**